internal/internal_workflow_client_test.go

```diff
@@ -976,6 +976,161 @@ func (s *workflowRunSuite) TestGetWorkflowNoExtantWorkflowAndNoRunId() {
 	s.Equal("", workflowRunNoRunID.GetRunID())
 }
 
+func (s *workflowRunSuite) TestExecuteWorkflowWithUpdate_InvalidUpdateWorkflowOptions() {
+	_, err := NewUpdateWorkflowOperation(
+		UpdateWorkflowOptions{
+			// invalid
+		})
+	s.ErrorContains(err, "WaitForStage must be specified")
+
+	_, err = NewUpdateWorkflowOperation(
+		UpdateWorkflowOptions{
+			WaitForStage: WorkflowUpdateStageCompleted,
+			RunID:        "invalid",
+		})
+	s.ErrorContains(err, "RunID is not allowed to be used on an UpdateWorkflowOperation")
+
+	_, err = NewUpdateWorkflowOperation(
+		UpdateWorkflowOptions{
+			WaitForStage:        WorkflowUpdateStageCompleted,
+			FirstExecutionRunID: "invalid",
+		})
+	s.ErrorContains(err, "FirstExecutionRunID is not allowed to be used on an UpdateWorkflowOperation")
+}
+
+func (s *workflowRunSuite) TestExecuteWorkflowWithUpdate_NonMultiOperationError() {
+	s.workflowServiceClient.EXPECT().
+		ExecuteMultiOperation(gomock.Any(), gomock.Any(), gomock.Any()).
+		Return(nil, serviceerror.NewInternal("internal error")).Times(1)
+
+	updOp, err := NewUpdateWorkflowOperation(
+		UpdateWorkflowOptions{
+			UpdateName:   "update",
+			WaitForStage: WorkflowUpdateStageCompleted,
+		})
+	s.Nil(err)
+
+	_, err = s.workflowClient.ExecuteWorkflow(
+		context.Background(),
+		StartWorkflowOptions{
+			ID:                 workflowID,
+			TaskQueue:          taskqueue,
+			WithStartOperation: updOp,
+		}, workflowType,
+	)
```

Comment:
@cretz @Quinn-With-Two-Ns 

With the naming discussion in mind, what do y'all think about this direction?

```go
op := client.UpdateWithStartWorkflowOperation{
	UpdateName:   "UPDATE",
	Args:         []interface{}{5},
	WaitForStage: client.WorkflowUpdateStageCompleted,
}

wfHandle, err := c.ExecuteWorkflow(
	context.Background(),
	client.StartWorkflowOptions{
		...
		WithStartOperation:       op,
	},
	Workflow,
)
if err != nil {
	panic(err)
}

var updateResult int
updHandle, err := op.Get(context.Background())
if err != nil {
	panic(err)
}
```

Benefits:

- removes need for constructor
- easier to find from option `WithStartOperation`'s type: just look for implementations in IDE
- since the operation is just a struct, there's no risk assuming it performs an action of any kind
- one less error handling check
- `UpdateWithStartWorkflowOperation` can be a copy of `UpdateWorkflowOptions` (and omit invalid fields)

Downsides:

- extra type to maintain / no re-use (to keep in sync with `UpdateWorkflowOptions`) [could also embed instead]
- validation error performed at execution stage, ie later

Reply:
This looks good to me. Assuming you are maintaining private state on that struct though, make sure to mention that 1) fields are not expected to be mutated after it is in use (and either have no effect or would would be racy), and 2) that the struct is only usable once.

Reply:
If we go with this approach and naming, I recommend renaming `Args` as `UpdateArgs` to make it clear _which_ args it is.

Reply:
Yeah there is a another downside as Chad pointed out that this lets the user modify struct field after call which is not valid and it is not consistent with the rest of the Go SDK. I prefer the current approach maybe just renaming `NewUpdateWorkflowOperation` to `NewUpdateWorkflowWithStartOperation` and take a `UpdateWorkflowWithStartOperationOptions` as a parameter.

Reply:
Notes/input from a meeting just now:

- it's advantageous to run the Update validation from `ExecuteWorkflow` since then the interceptors are in the loop
- using a function instead of a struct here is more idiomatic Go
- UpdateWithStart works without "workflow" suffix as it's a proper name (will be used through the product/docs)
- using the existing `UpdateWorkflowOptions` makes it easier for users to migrate from Update to Update-with-Start, but risks that users use invalid parameters only to find out at runtime - whereas using a dedicated struct makes migrating harder but is safer

Reply:
> UpdateWithStart works without "workflow" suffix as it's a proper name (will be used through the product/docs)

For the same reasons we call it `ExecuteWorkflow` and not just `Execute`, and `UpdateWorkflow` and not `Update`, I think it's important to use the word workflow when on a generic thing or top-level and talking about something specific to a workflow. We do it everywhere else and I don't think it's worth saving a word to become inconsistent in this special case.